text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,316 @@
+
+    ```rust
+    enum E { Var1, Var2 }
+
+    const BEST_VAR: E = E::Var1;
+
+    fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
+    fn is_var1(e: E) -> bool { matches!(e, E::Var1) }
+    ```
+
+    It would be very surprising if those two functions could behave differently.
+  It follows that we cannot allow `PartialEq` to affect the behavior of constants in patterns.
+
+- We do not want to expose equality tests on types where the library author did not explicitly expose an equality test.
+  This means not allowing matching on constants whose type does not implement `PartialEq`.
+  It also means not allowing matching on constants where running `PartialEq` would behave different from the corresponding pattern:
+  the pattern can be accessing private fields, and the `PartialEq` implementation provided by the library might be treating those fields in a particular way;
+  we should not let people write patterns that "bypass" any such treatment and do structural matching when the crate author does not provide a structural `PartialEq`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Constants can be used as patterns, but only if their type implements `PartialEq`.
+Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):
+
+```rust
+#[derive(PartialEq)] // code fails to build if we remove this or replace it by a manual impl
+enum E { Var1, Var2 }
+
+const BEST_VAR: E = E::VAR1;
+
+fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
+```
+
+```rust
+#[derive(PartialEq)]
+struct S { f1: i32, f2: E }
+
+const MY_S: S = S { f1: 42, f2: BEST_VAR };
+
+// Removing *either* `derive` or implementing `PartialEq` manually would lead to rejecting this code
+fn is_mine(s: S) -> bool { matches!(s, MY_S) }
+```
+
+We say that a type that derives `PartialEq` has "structural equality": the
+equality of this type is defined fully by the equality on its fields.
+
+For matching on values of enum type, it is sufficient if the actually chosen
+variant has structural equality; other variants do not matter:
+
+```rust
+struct NonStructuralEq(i32);
+
+impl PartialEq for NonStructuralEq {
+    fn eq(&self) -> bool { true }
+}
+
+#[derive(PartialEq)]
+enum MyEnum { GoodVariant(i32), NonStructuralVariant(NonStructuralEq) }
+
+// This constant *can* be used in a pattern.
+const C: MyEnum = MyEnum::GoodVariant(0);
+```
+
+This means the eligibility of a constant for a pattern depends on its value, not just on its type.
+That is already the case on stable Rust for many years and relied upon by widely-used crates such as [`http`](https://github.com/rust-lang/rust/issues/62411#issuecomment-510604193).
+
+Overall we say that the *value* of the constant must have recursive structural equality,
+which is the case when all the types that actually appear recursively in the type (ignoring "other" enum variants) have structural equality.
```

Comment:
Is `derive(PartialEq)` necessary on the enum, or just on the contents of the enum variant that's used in the constant?

I think it's required, but saying "appear recursively in the type of the value" is potentially ambiguous. 

```suggestion
which is the case when all the types that actually appear recursively in the type (ignoring "other" enum variants) have structural equality. For a variant to have structural equality, the enum must derive `PartialEq`.
```

Reply:
Ah I see this is explained in the reference section. Perhaps this clarification isn't needed then.